component/module/module.go

```diff
@@ -0,0 +1,131 @@
+package module
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"path"
+	"sync"
+	"time"
+
+	"github.com/gorilla/mux"
+	"github.com/grafana/agent/component"
+	"github.com/grafana/agent/pkg/flow"
+	"github.com/grafana/agent/pkg/flow/logging"
+	"github.com/grafana/agent/pkg/flow/tracing"
+	"github.com/grafana/agent/web/api"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+// ModuleComponent holds the common properties for module components.
+type ModuleComponent struct {
+	opts component.Options
+	ctrl *flow.Flow
+
+	mut    sync.RWMutex
+	health component.Health
+}
+
+// Exports holds values which are exported from the run module.
+type Exports struct {
+	// Exports exported from the running module.
+	Exports map[string]any `river:"exports,attr"`
+}
+
+// NewModuleComponent initializes a new ModuleComponent.
+func NewModuleComponent(o component.Options) ModuleComponent {
+	// TODO(rfratto): replace these with a tracer/registry which properly
+	// propagates data back to the parent.
+	flowTracer, _ := tracing.New(tracing.DefaultOptions)
+	flowRegistry := prometheus.NewRegistry()
+
+	return ModuleComponent{
+		opts: o,
+		ctrl: flow.New(flow.Options{
+			ControllerID: o.ID,
+			LogSink:      logging.LoggerSink(o.Logger),
+			Tracer:       flowTracer,
+			Reg:          flowRegistry,
+
+			DataPath:       o.DataPath,
+			HTTPPathPrefix: o.HTTPPath,
+			HTTPListenAddr: o.HTTPListenAddr,
+
+			OnExportsChange: func(exports map[string]any) {
+				o.OnStateChange(Exports{Exports: exports})
+			},
+		}),
+	}
+}
+
+// LoadFlowContent loads the flow controller with the current component content.
+func (c *ModuleComponent) LoadFlowContent(arguments map[string]any, contentValue string) error {
+	c.mut.RLock()
```

Comment:
I wonder if its clearer to wrap all this in a Rlock with defer instead of doing it in sections.

Reply:
Good question, that's something I've been wondering about in go. Should I try to hold locks for the absolute minimal amount of time or value the clarity of the defer pattern when it's 'close enough'?

Reply:
locks removed. I used to lock for the content/arguments but those are now function params after the refactor